[rtextures] `HalfToFloat()` and `FloatToHalf()` dereferencing issues

@Not-Nik @raysan5 Sorry for raising the PR again, but while compiling raylib with `-02` (or `-03`) `HalfToFloat` and `FloatToHalf` will raise `dereferencing type-punned pointer will break strict-aliasing rules`, which, AFAIK, incurs in undefined behavior, so probably worth looking into. Not sure how to address these ones tho.
EDIT: `-fno-strict-aliasing` is set, not sure why it's not being applied.

<details><summary>Compiling log:</summary>

``` bash
[...]
gcc -c rtextures.c -Wall -D_GNU_SOURCE -DPLATFORM_DESKTOP_GLFW -DGRAPHICS_API_OPENGL_33 -Wno-missing-braces -Werror=pointer-arith -fno-strict-aliasing -std=c99 -fPIC -O2 -Werror=implicit-function-declaration -D_GLFW_X11  -I.  -Iexternal/glfw/include
rtextures.c: In function ‘HalfToFloat’:
rtextures.c:5396:30: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
 5396 |     const unsigned int v = (*(unsigned int *)&fm) >> 23; // Evil log2 bit hack to count leading zeros in denormalized format
      |                              ^~~~~~~~~~~~~~~~~~~
rtextures.c:5399:15: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
 5399 |     result = *(float *)&r;
      |               ^~~~~~~~~~~
rtextures.c: In function ‘FloatToHalf’:
rtextures.c:5409:30: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
 5409 |     const unsigned int b = (*(unsigned int *) & x) + 0x00001000; // Round-to-nearest-even: add last bit after truncated mantissa
      |                              ^~~~~~~~~~~~~~~~~~~~
rtextures.c: In function ‘HalfToFloat’:
rtextures.c:5396:29: warning: ‘fm’ is used uninitialized [-Wuninitialized]
 5396 |     const unsigned int v = (*(unsigned int *)&fm) >> 23; // Evil log2 bit hack to count leading zeros in denormalized format
      |                            ~^~~~~~~~~~~~~~~~~~~~~
rtextures.c:5395:17: note: ‘fm’ declared here
 5395 |     const float fm = (float)m;
      |                 ^~
[...]
```
</details>

_Originally posted by @asdqwe in https://github.com/raysan5/raylib/issues/3220#issuecomment-2614449236_

--- a/src/rtextures.c
+++ b/src/rtextures.c
@@ -5396,7 +5396,6 @@ static float HalfToFloat(unsigned short x)
     float result = 0.0f;
 
     union floatUnsignedUnion uni;
-    uni.fm = 0.0f;
 
     const unsigned int e = (x & 0x7C00) >> 10; // Exponent
     const unsigned int m = (x & 0x03FF) << 13; // Mantissa

--- a/src/rtextures.c
+++ b/src/rtextures.c
@@ -5384,18 +5384,17 @@ int GetPixelDataSize(int width, int height, int format)
 //----------------------------------------------------------------------------------
 // Module specific Functions Definition
 //----------------------------------------------------------------------------------
-union floatUnsignedUnion {
-    float fm;
-    unsigned int ui;
-};
-
 // Convert half-float (stored as unsigned short) to float
 // REF: https://stackoverflow.com/questions/1659440/32-bit-to-16-bit-floating-point-conversion/60047308#60047308
 static float HalfToFloat(unsigned short x)
 {
     float result = 0.0f;
 
-    union floatUnsignedUnion uni;
+    union
+    {
+        float fm;
+        unsigned int ui;
+    } uni;
 
     const unsigned int e = (x & 0x7C00) >> 10; // Exponent
     const unsigned int m = (x & 0x03FF) << 13; // Mantissa
@@ -5413,7 +5412,11 @@ static unsigned short FloatToHalf(float x)
 {
     unsigned short result = 0;
 
-    union floatUnsignedUnion uni;
+    union
+    {
+        float fm;
+        unsigned int ui;
+    } uni;
     uni.fm = x;
 
     const unsigned int b = uni.ui + 0x00001000; // Round-to-nearest-even: add last bit after truncated mantissa

--- a/src/rtextures.c
+++ b/src/rtextures.c
@@ -5384,19 +5384,27 @@ int GetPixelDataSize(int width, int height, int format)
 //----------------------------------------------------------------------------------
 // Module specific Functions Definition
 //----------------------------------------------------------------------------------
+union floatUnsignedUnion {
+    float fm;
+    unsigned int ui;
+};
+
 // Convert half-float (stored as unsigned short) to float
 // REF: https://stackoverflow.com/questions/1659440/32-bit-to-16-bit-floating-point-conversion/60047308#60047308
 static float HalfToFloat(unsigned short x)
 {
     float result = 0.0f;
 
+    union floatUnsignedUnion uni;
+    uni.fm = 0.0f;
+
     const unsigned int e = (x & 0x7C00) >> 10; // Exponent
     const unsigned int m = (x & 0x03FF) << 13; // Mantissa
-    const float fm = (float)m;
-    const unsigned int v = (*(unsigned int *)&fm) >> 23; // Evil log2 bit hack to count leading zeros in denormalized format
-    const unsigned int r = (x & 0x8000) << 16 | (e != 0)*((e + 112) << 23 | m) | ((e == 0)&(m != 0))*((v - 37) << 23 | ((m << (150 - v)) & 0x007FE000)); // sign : normalized : denormalized
+    uni.fm = (float)m;
+    const unsigned int v = uni.ui >> 23; // Evil log2 bit hack to count leading zeros in denormalized format
+    uni.ui = (x & 0x8000) << 16 | (e != 0)*((e + 112) << 23 | m) | ((e == 0)&(m != 0))*((v - 37) << 23 | ((m << (150 - v)) & 0x007FE000)); // sign : normalized : denormalized
 
-    result = *(float *)&r;
+    result = uni.fm;
 
     return result;
 }
@@ -5406,7 +5414,10 @@ static unsigned short FloatToHalf(float x)
 {
     unsigned short result = 0;
 
-    const unsigned int b = (*(unsigned int *) & x) + 0x00001000; // Round-to-nearest-even: add last bit after truncated mantissa
+    union floatUnsignedUnion uni;
+    uni.fm = x;
+
+    const unsigned int b = uni.ui + 0x00001000; // Round-to-nearest-even: add last bit after truncated mantissa
     const unsigned int e = (b & 0x7F800000) >> 23; // Exponent
     const unsigned int m = b & 0x007FFFFF; // Mantissa; in line below: 0x007FF000 = 0x00800000-0x00001000 = decimal indicator flag - initial rounding
 

